src/main/java/sparksoniq/jsoniq/item/ItemFactory.java

```diff
@@ -25,50 +25,48 @@ public static ItemFactory getInstance() {
         }
         return _instance;
     }
-    
-    public Item createStringItem(String s) {
+
+    public StringItem createStringItem(String s) {
```

Comment:
Taking _createStringItem(String s)_ as example: 

I feel like returning _Item_ instead of _StringItem_ would then potentially necessitate for checks like _isString()_ or instance-of, and potentially to cast to the StringItem subclass (since we are losing information on which type of Item we are dealing with). On the other hand, returning a _StringItem_, would allow for the return value to be used both as Item and StringItem (e.g. `Item item = createStringItem("hello")` and also `StringItem item = createStringItem("hello")`, that is not possible returning an object of type _Item_. 

At the same time, a if we return an object of type _StringItem_, it can be passed as parameter to a function that takes a _StringItem_ as argument, but not if we return an object of type _Item_, unless we perform a check _isString()_ or instance-of check, or cast (e.g. `functionThatTakesStringItem(createStringItem("hello"))` is fine if _createStringItem(String s)_ returns a _StringItem_, not if it returns an _Item_ .

Regarding the concern on keeping the internal implementation as black box to the external facing API, these functions are in the _ItemFactory_ class which (so far) is internal.

I do not mind keeping every return type to _Item_ (I already reverted my changes), but I do not really see the advantage. On the contrary, I feel like it might be an option that potentially requires more checks and casts.

Reply:
Thank you for reverting and for your understanding.

I understand your point from an OO-perspective -- it makes perfect sense and in general, you have a strength in write good OO code.

This needs to be looked at, here, from the perspective of performance: code that gets executed billions of times in loops must avoid as much as possible instance-of tests, casts as well as creation of new objects, which are expensive.

Instance-of tests are avoided by using is...() function calls instead. Casts are avoided by using the superclass in all signatures. Construction of objects is avoided by pooling or sharing instances for entire loops.

The reason that, in the design, functions in general all take an Item is that no cast is ever needed, also in case the input item (for example output by an iterator, which happens often) is an Item. A function that would take a StringItem as an argument forces such casts, which has an impact on performance. This is why we avoid doing so (even though it would have been very good practice for software where performance is not critical).

The price to pay for this is indeed that additional checks may be required -- this is a compromise. But again from the perspective of performance, this sort of check is not very expensive and only takes a few CPU cycles, which is the most important for this sort of software.

As a rule of thumb: 1. correctness (tests) first 2. then performance 3. then readability by other people in the team who may have different coding mindsets and 4. then OO design.

Does it makes sense?

Reply:
Yes, very clear.
Thanks for the explanation!